processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingOptions.java

```diff
@@ -182,6 +195,7 @@ public boolean hasNestedTargetReferences() {
     public Map<Parameter, MappingOptions> groupBySourceParameter() {
 
         Map<Parameter, List<Mapping>> mappingsKeyedByParameterType = new HashMap<Parameter, List<Mapping>>();
```

Comment:
`mappingsKeyedByParameterType` = `mappingsKeyedByParameter`. This is deliberate. Look at the equals in `Parameter` and what makes Parameter unique..